### PR TITLE
filter_chain: refactor code

### PR DESCRIPTION
**Pull request by kyrylo**
* filter_chain: rename 'keys_filters' to 'post_filters' 
* filter_chain: sort filters by weight

### CODE REVIEW

**Comment by vmihailenco** [comment]
I never saw prelude+postlude names in APIs. Usually before/after & pre&post names are used.

Overall you have prelude & postlude but nothing in between which does not make much sense.

File: lib/airbrake-ruby/filter_chain.rb [hunk 1]
```diff
@@ -45,11 +45,11 @@ def initialize(config)
...
     def add_filter(filter)
-      return @keys_filters << filter if KEYS_FILTERS.include?(filter.class)
-      return @filters << filter unless filter.to_s.start_with?(LIB_NAMESPACE)
+      return @postlude_filters << filter if POSTLUDE_FILTERS.include?(filter.class)
+      return @prelude_filters << filter unless filter.to_s.start_with?(LIB_NAMESPACE)
 
-      i = @filters.rindex { |f| f.to_s.start_with?(LIB_NAMESPACE) }
-      @filters.insert(i + 1, filter) if i
+      i = @prelude_filters.rindex { |f| f.to_s.start_with?(LIB_NAMESPACE) }
+      @prelude_filters.insert(i + 1, filter) if i
```
Comment (vmihailenco) [inline lib/airbrake-ruby/filter_chain.rb]: This function looks unnecessary cryptic.

**Comment by kyrylo** [comment]
Let's go with `before/after`. How about leaving `@filters` but renaming `@keys_filters` to `@post_filters`?

File: lib/airbrake-ruby/filter_chain.rb [hunk 2]
```diff
@@ -59,7 +59,7 @@ def add_filter(filter)
     # @param [Airbrake::Notice] notice The notice to be filtered
     # @return [void]
     def refine(notice)
-      (@filters + @keys_filters).each do |filter|
+      (@prelude_filters + @postlude_filters).each do |filter|
```
Comment (vmihailenco) [inline lib/airbrake-ruby/filter_chain.rb]: I guess this creates new array each time notice is filtered?

**Comment by vmihailenco** [comment]
>How about leaving @filters but renaming @keys_filters to @post_filters?

Sounds better.

File: lib/airbrake-ruby/filter_chain.rb (hunk 2 above)
Comment (kyrylo) [inline lib/airbrake-ruby/filter_chain.rb]: I left only one array. We now `insert` and calculate indices.

File: lib/airbrake-ruby/filter_chain.rb (hunk 1 above)
Comment (kyrylo) [inline lib/airbrake-ruby/filter_chain.rb]: Refactored.

**Comment by kyrylo** [comment]
PTAL

File: lib/airbrake-ruby/filter_chain.rb [hunk 3]
```diff
@@ -45,11 +54,12 @@ def initialize(config)
...
-      return @filters << filter unless filter.to_s.start_with?(LIB_NAMESPACE)
-
-      i = @filters.rindex { |f| f.to_s.start_with?(LIB_NAMESPACE) }
-      @filters.insert(i + 1, filter) if i
+      if filter.to_s.start_with?(LIB_NAMESPACE)
+        @filters.insert(@offset_left, filter)
+        @offset_left += 1
+      else
+        @filters.insert(@filters.size - @offset_right, filter)
+      end
```
Comment (vmihailenco) [inline lib/airbrake-ruby/filter_chain.rb]: This is still ugly and unreadable. What is the problem with adding optional `weight` property to filters (with default value 0) and sort by `filter.weight`? It should not be more than 10 lines of code.

Comment (kyrylo) [inline lib/airbrake-ruby/filter_chain.rb]: Sure, let's try. I don't think there's a problem. Filter weights will give a lot of flexibility, but nobody has requested this flexibility so far.

**Comment by kyrylo** [comment]
All library filters have a weight of 0. The blacklist/whitelist filters have a weight of 2. User defined filters will have a weight of 1.

PTAL:

**Comment by kyrylo** [comment]
I'm also not documenting this because nobody has requested the feature. So far it's purely for internal use.

**Comment by vmihailenco** [comment]
Generally LGTM

There are 2 moderately confusing things:
- filters with bigger weight are run later but I expect the opposite
- the default is 1, but I expect 0. It is especially confusing to see that you set optional field `weight = 0`.

**Comment by kyrylo** [comment]
If the default is 0, then keys filters will need to have a weight of `-1`, user filters will be `0` (default), library filters will be `1`. Sounds good?

**Comment by vmihailenco** [comment]
Yes, except that I will go with -100, 0, 100 to give some room to insert filters in between. You can go even further and use `-100, -90, 0 (default), 100, 110, 120`, i.e. don't use same weight twice.

**Pull request merged**